File: 199. Binary Tree Right Side View/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    
    var depth []int
    var stack []*TreeNode
    var view []int
    var d int
    var D int
    view = append(view, root.Val)
    stack, depth = push(stack, root.Left, depth, d+1)
    node := root.Right
    d++
    
    for len(stack) != 0 || node != nil {
        if node == nil {
            stack, node, depth, d = pop(stack, depth)
        } else {
            if d > D {
                view = append(view, node.Val)
                D++
            }
            if node.Left != nil {
                stack, depth = push(stack, node.Left, depth, d+1)
            }
            node = node.Right
            d++
        }
    }
    return view
}

func push(stack []*TreeNode, node *TreeNode, depth []int, d int) ([]*TreeNode, []int) {
    stack = append(stack, node)
    depth = append(depth, d)
    return stack, depth
}

func pop(stack []*TreeNode, depth []int) ([]*TreeNode, *TreeNode, []int, int) {
    l := len(stack)-1
    node := stack[l]
    stack = stack[:l]
    d := depth[l]
    depth = depth[:l]
    return stack, node, depth, d
}

File: 199. Binary Tree Right Side View/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        self.myRightSideView(root, 0, res)
        return res
        
    def myRightSideView(self, root, depth, response):
        if not root:
            return
        
        if depth == len(response):
            response.append(root.val)
        
        self.myRightSideView(root.right, depth + 1, response)
        self.myRightSideView(root.left, depth + 1, response)
        
